Add blog post: GDB support inside Gramine's SGX enclaves

pwmarcz opened the pull request:
This includes a "Blog" section, and a first blog post.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)
<!-- Reviewable:end -->

pwmarcz committed:
Add blog section

Signed-off-by: Paweł Marczewski <pwmarcz@gmail.com>

pwmarcz committed:
Add blog post: GDB support inside Gramine's SGX enclaves

Signed-off-by: Paweł Marczewski <pwmarcz@gmail.com>

pwmarcz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: 0 of 14 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv, @pwmarcz, and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N06R-jI16iP9Itd85qj:-N06R-jI16iP9Itd85qk:b-iqz2nd) (no related file):*
FYI: I suggest viewing the rendered blog post (see README).

---
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 5 at r1](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N06REx1-7KPTyPsE76r:-N06REx1-7KPTyPsE76s:b-s2btbh) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/7fa97861dd53ae7a5219185fe6da81b3131919f0/content/blog/gdb-support-in-gramine/contents.lr#L5)):*
> ```Markdown
> author: Paweł Marczewski
> ---
> pub_date: 2022-04-20
> ```

TODO on merge: update date
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 17 at r1](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N06RBtyAO5W7--RGsls:-N06RBtyAO5W7--RGslt:b-4xxvyb) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/7fa97861dd53ae7a5219185fe6da81b3131919f0/content/blog/gdb-support-in-gramine/contents.lr#L17)):*
> ```Markdown
> *This is a blog post about GDB support in the Gramine project. This feature was
> originally written by Chia-Che Tsai, and later expanded by me (Paweł Marczewski)
> and other contributors to the project. Thanks to Michał Kowalczyk for reviewing
> ```

TODO on merge: add reviewers


<!-- Sent from Reviewable.io -->

woju reviewed (changes requested):
> Reviewed 8 of 14 files at r1.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: 8 of 14 files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @pwmarcz)

___
*[content/blog/contents.lr, line 1 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N074eIO1IIZgB7t6arz:-N074eIO1IIZgB7t6as-:b1ruf2k) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/contents.lr#L1)):*
> ```Markdown
> 
> ```

I'd prefer explicit `_model: blog` instead of relying on [automatic model selection](https://www.getlektor.com/docs/models/selection/) here (autoselection in children is fine, because it's specified in parent model). Alternatively, if you're sure autoselection here is fine, why is there a single `\n` byte, and this isn't just an empty file?
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 3 at r1](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N07B3NT25_sBhkT8Brn:-N07B3NT25_sBhkT8Bro:b-f5aa2m) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/7fa97861dd53ae7a5219185fe6da81b3131919f0/content/blog/gdb-support-in-gramine/contents.lr#L3)):*
> ł

Just noticed, the font chosen by designers lacks Polish characters (tbh it has only english and arabic) and by a looks of it, it's only a repackage of Source Sans Pro. Do we have any arabic text anywhere? If not, can you add another commit that would switch "Mada" to "Source Sans Pro"? (Currently we load from google fonts, so this should be easy).


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> Reviewed 13 of 14 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: all files reviewed, 19 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @pwmarcz)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09oZOy6Y5iqGHp-vFO:-N09oZOy6Y5iqGHp-vFP:b-n32rqs) (no related file):*
I don't like the black background, but I guess we don't want to change the colors only for the blog.

---
*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wrvTAwgH08Fafoio:-N09wrvTAwgH08Fafoip:b6hg7gc) (no related file):*
The "Reality vs What GDB sees" diagram is a bit too long -- you have a lot of unused space at the bottom, around `Untrusted runtime` boxes. You can easily decrease the height of this diagram.

---
*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09y1Hp2l_n93lFmIK6:-N09y1Hp2l_n93lFmIK7:brmajod) (no related file):*
The blog post looks cool, thanks for this work!

---
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 20 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09waCa03s1Bv5CFDNi:-N09waCa03s1Bv5CFDNj:b-g7abuj) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L20)):*
> ```Markdown
> the post.*
> 
> [Gramine] is a tool for running Linux applications under non-standard
> ```

`tool` -> `framework`? Or maybe `runtime`?
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 33 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wBvOEMJhjj3xWJKH:-N09wBvOEMJhjj3xWJKI:b-as2u49) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L33)):*
> ```Markdown
> external world (for example remote hosts), but it still trusts the host OS when
> it comes to basic operations like saving data in files, or sending it between
> processes. If we do not want the host to access (or modify) our data, performing
> ```

I would start the sentence with `In the case of SGX however, we do not want the host ..., so performing these operations ...`
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 36 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wKG-7LohZgHNhMBR:-N09wKG-7LohZgHNhMBS:b-k9zl8a) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L36)):*
> ```Markdown
> these operations will be more complicated than just asking the host OS.
> 
> Instead, Gramine itself acts as an operating system, and handles requests
> ```

`Instead`? This is ambigious, instead of what? Maybe `To protect the application from malicious host, Gramine ...`
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 36 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wW3596UZw6DJooin:-N09wW36A1l5XipVS_4V:b-exti1b) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L36)):*
> ```Markdown
> these operations will be more complicated than just asking the host OS.
> 
> Instead, Gramine itself acts as an operating system, and handles requests
> ```

`handles` -> `intercepts`?
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 55 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09vm-Y4zv5f_O-J7Pd:-N09vm-Y4zv5f_O-J7Pe:b-bs1egi) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L55)):*
> ```Markdown
> get GDB running, but it wasn't easy...
> 
> ## Looking inside
> ```

Maybe `Looking inside the SGX enclave`?
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 61 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09xW562f2335DBTIDI:-N09xW562f2335DBTIDJ:b-xkrjtc) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L61)):*
> ```Markdown
> 
> For debugging purposes, the rules can be relaxed somewhat: if the enclave runs
> with a special `DBGOPTIN` flag, direct memory access will still not work, but
> ```

It's a bit more complicated. Two things are needed:
- The enclave itself must be a `debug` enclave (bit 1 in SECS.ATTRIBUTES)
- Enclave TCS (= one enclave thread) must set this `DBGOPTIN` flag in its TCS.FLAGS.

I don't know if you want to add these details though. Maybe just mention that `the enclave must have a debug bit set plus all enclave threads must set the DBGOPTIN flag`.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 70 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09x7pWDHvbxT9Y5SyN:-N09x7pWDHvbxT9Y5SyO:b1yuxds) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L70)):*
> ```Markdown
> 
> Unfortunately, it's not that easy to check which part of the program is being
> executed. When the process is stopped, it automatically exits from the enclave.
> ```

`it automatically exists` sounds weird. I'd rephrase to `the SGX hardware forces the process to exit the enclave and land on the so-called Async Exit Pointer (AEP)`.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 95 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09uE9l8UVxkkYSTOm4:-N09uE9l8UVxkkYSTOm5:b-ug9569) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L95)):*
> ```Markdown
> Using this wrapper, we can ensure that GDB sees the right register values.
> Whenever we find the process stopped at the AEP, we read the "real" instruction
> pointer and other registers through `/proc/<pid>/mem`, and return this
> ```

Actually, how do we read the "real" instruction pointer? How do we know where this instruction pointer is?

The answer is a bit complicated, because it requires deep knowledge of SGX hardware structures (in particular, the layout of TCS and SSA frames). But basically Gramine dumps info about all TCSes in a special out-of-enclave struct: https://github.com/gramineproject/gramine/blob/ad85fe95c43fd583f7fa53a5f170383a5fcd8536/Pal/src/host/Linux-SGX/gdb_integration/sgx_gdb.h#L16.

Then our GDB wrapper consults this struct, extracts the base TCS address for a current thread, calculates the SSA frame's base address and calculates the address where GPRs + XSAVE area are located. Then our GDB wrapper reads them through `/proc/pid/mem` (which internally just invokes `EDBGRD`/`EDBGWR` instructions), and voila, all register values are known.

I think this is an important detail that is missing here. Again, I understand that this is all super low-level, but at least some links to the relevant sources will help a curious reader.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 100 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09q1aT2HE8L1Ld4LX0:-N09q1aT2HE8L1Ld4LX1:b81f228) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L100)):*
> ```Markdown
> 
> [ld_preload]: https://stackoverflow.com/questions/426230/what-is-the-ld-preload-trick
> [ptrace_wrapper]: (https://github.com/gramineproject/gramine/blob/ad85fe95c43fd583f7fa53a5f170383a5fcd8536/Pal/src/host/Linux-SGX/gdb_integration/sgx_gdb.c#L488)
> ```

I guess you don't need `()` here -- my Lektor render shows this link incorrectly.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 132 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09sXc79otP-8oxkG_C:-N09sXc79otP-8oxkG_D:b958hi9) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L132)):*
> ```Markdown
> implemented a similar solution ourselves: we maintain [our own
> structure][debug_map] inside Gramine, and read it in a [Python extension for
> GDB][debug_map_py].
> ```

Some details are missing here. You don't mention that our in-enclave patched Glibc (well, more technically the `ld` library) introduces a special call into Gramine -- `gramine_register_library()` -- during binary loading. And of course during `execve()` syscall emulation. And only in this special call, Gramine calls a special `debug_map_update_debugger()` which is hooked into GDB by our integration script.

I find these details fascinating and worth describing, even if very briefly -- it gives at least some hints for interested devs, where to look for corresponding source code.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 169 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09r4Ma9vaWJoOwTN1V:-N09r4Mb--8n8izeAh2_:b7sm4ru) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L169)):*
> ```Markdown
> 
> As a result, if we want to read a full stack trace, we might need to go through
> three different places in memory:
> ```

You have this sentence twice, please remove one copy.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 171 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09rQ7m6y7pIouE20ep:-N09rQ7m6y7pIouE20eq:b-mct0ls) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L171)):*
> ```Markdown
> three different places in memory:
> 
> <img src="/blog/gdb-support-in-gramine/stack.svg" style="margin-top: -50px; margin-bottom: -50px; filter: invert(100%);">
> ```

In the picture, you have `write(1, "hello")` twice: the upper-most stack frame is for the untrusted runtime, the lower stack frame is for the trusted in-enclave runtime.

I suggest to give some hint, e.g. adding ` (host-level syscall)` and ` (enclave-level syscall)` after each of these two, to distinguish them.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 178 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09ryTa9GjkP2QVoQzT:-N09ryTa9GjkP2QVoQzU:b-1kuavi) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L178)):*
> ```Markdown
> saved at address `[RSP+xyz]`").
> 
> [CFI directives]: https://github.com/gramineproject/gramine/blob/e37f773d4d910799a854b90334f2ff8eb6f411bd/Pal/src/host/Linux-common/debug_map.c#L282
> ```

I find this link strange. I thought you wanted to give some real-world examples in this link, or at least point to some GDB official documentation.


<!-- Sent from Reviewable.io -->

pwmarcz committed:
Change font family to Source Sans Pro

This fixes rendering of characters such as Polish 'ł', not included in
the Mada font family used before.

(Mada is based on Source Sans Pro, with an additional Arabic component.
As far as I can tell, there's no Arabic text on the website.)

Signed-off-by: Paweł Marczewski <pwmarcz@gmail.com>

pwmarcz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: 9 of 14 files reviewed, 19 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv, @pwmarcz, and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09oZOy6Y5iqGHp-vFO:-N0As5Km4D7Ds_kcgEx7:b-8fdq9g) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I don't like the black background, but I guess we don't want to change the colors only for the blog.
</blockquote></details>

Yeah, I think a light background is doable, but I don't really want to work on frontend right now :)

---
*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wrvTAwgH08Fafoio:-N0At7IR0RitSEpLstam:b-bw8e14) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

The "Reality vs What GDB sees" diagram is a bit too long -- you have a lot of unused space at the bottom, around `Untrusted runtime` boxes. You can easily decrease the height of this diagram.
</blockquote></details>

Updated.

---
*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09y1Hp2l_n93lFmIK6:-N0At9In2eU00Sdk1-l3:b-i8euqt) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

The blog post looks cool, thanks for this work!
</blockquote></details>

Thanks. If all goes well, I want to write one or two more (about SGX profiling, and ASan).

---
___
*[content/blog/contents.lr, line 1 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N074eIO1IIZgB7t6arz:-N0ATHyu5uJ_r8t5_mKZ:bxwmib) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/contents.lr#L1)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

I'd prefer explicit `_model: blog` instead of relying on [automatic model selection](https://www.getlektor.com/docs/models/selection/) here (autoselection in children is fine, because it's specified in parent model). Alternatively, if you're sure autoselection here is fine, why is there a single `\n` byte, and this isn't just an empty file?
</blockquote></details>

Makes sense. Updated.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 3 at r1](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N07B3NT25_sBhkT8Brn:-N0ATbdu74cE0bj2oEVo:bkjdnwu) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/7fa97861dd53ae7a5219185fe6da81b3131919f0/content/blog/gdb-support-in-gramine/contents.lr#L3)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

> ł

Just noticed, the font chosen by designers lacks Polish characters (tbh it has only english and arabic) and by a looks of it, it's only a repackage of Source Sans Pro. Do we have any arabic text anywhere? If not, can you add another commit that would switch "Mada" to "Source Sans Pro"? (Currently we load from google fonts, so this should be easy).
</blockquote></details>

I don't think we have Arabic anywhere. 

I changed the font as suggested, and our `ł`s look better now.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 20 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09waCa03s1Bv5CFDNi:-N0AWC8S9oqJQP3fCzh6:b8i9398) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L20)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`tool` -> `framework`? Or maybe `runtime`?
</blockquote></details>

Changed to "framework".
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 33 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wBvOEMJhjj3xWJKH:-N0AWQ3X8IB2PtzX8i9X:b-896fix) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L33)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I would start the sentence with `In the case of SGX however, we do not want the host ..., so performing these operations ...`
</blockquote></details>

Done.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 36 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wKG-7LohZgHNhMBR:-N0AWQh1EqrvrLxROFpI:b-yhlv50) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L36)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`Instead`? This is ambigious, instead of what? Maybe `To protect the application from malicious host, Gramine ...`
</blockquote></details>

"Instead" was referring to the previous sentence (i.e. more than "just asking the host OS"). I rephrased that, though.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 36 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09wW3596UZw6DJooin:-N0AX_l_6pIdS5ijktTK:b2nhktp) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L36)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`handles` -> `intercepts`?
</blockquote></details>

OK.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 55 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09vm-Y4zv5f_O-J7Pd:-N0AXjDGDQy_SZpgVkBA:b-896fix) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L55)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Maybe `Looking inside the SGX enclave`?
</blockquote></details>

Done.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 61 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09xW562f2335DBTIDI:-N0AYrNt7-6gWTuDYD-I:b-y09cga) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L61)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

It's a bit more complicated. Two things are needed:
- The enclave itself must be a `debug` enclave (bit 1 in SECS.ATTRIBUTES)
- Enclave TCS (= one enclave thread) must set this `DBGOPTIN` flag in its TCS.FLAGS.

I don't know if you want to add these details though. Maybe just mention that `the enclave must have a debug bit set plus all enclave threads must set the DBGOPTIN flag`.
</blockquote></details>

I reworded it, is it better now?

(No mention of "from another process" anymore, but it should follow from the mention of `/proc/<pid>/mem`).
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 70 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09x7pWDHvbxT9Y5SyN:-N0AarpD71nIhzbLsvB5:b-qj92pf) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L70)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`it automatically exists` sounds weird. I'd rephrase to `the SGX hardware forces the process to exit the enclave and land on the so-called Async Exit Pointer (AEP)`.
</blockquote></details>

Reworded.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 95 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09uE9l8UVxkkYSTOm4:-N0Ab4Y0DPIbQbq0hn0G:b-qyzuiz) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L95)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Actually, how do we read the "real" instruction pointer? How do we know where this instruction pointer is?

The answer is a bit complicated, because it requires deep knowledge of SGX hardware structures (in particular, the layout of TCS and SSA frames). But basically Gramine dumps info about all TCSes in a special out-of-enclave struct: https://github.com/gramineproject/gramine/blob/ad85fe95c43fd583f7fa53a5f170383a5fcd8536/Pal/src/host/Linux-SGX/gdb_integration/sgx_gdb.h#L16.

Then our GDB wrapper consults this struct, extracts the base TCS address for a current thread, calculates the SSA frame's base address and calculates the address where GPRs + XSAVE area are located. Then our GDB wrapper reads them through `/proc/pid/mem` (which internally just invokes `EDBGRD`/`EDBGWR` instructions), and voila, all register values are known.

I think this is an important detail that is missing here. Again, I understand that this is all super low-level, but at least some links to the relevant sources will help a curious reader.
</blockquote></details>

Yeah, it's a balance - the detail is interesting, but I'm afraid too much will be again boring for the reader. 

I added another paragraph of explanation (and the link to `enclave_dbginfo`), what do you think?
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 100 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09q1aT2HE8L1Ld4LX0:-N0AhKtP0HCCkByhq5lq:bidrs83) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L100)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I guess you don't need `()` here -- my Lektor render shows this link incorrectly.
</blockquote></details>

Right. Fixed.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 132 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09sXc79otP-8oxkG_C:-N0AoOIn1DC1GMBXXooX:bs4p4g8) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L132)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Some details are missing here. You don't mention that our in-enclave patched Glibc (well, more technically the `ld` library) introduces a special call into Gramine -- `gramine_register_library()` -- during binary loading. And of course during `execve()` syscall emulation. And only in this special call, Gramine calls a special `debug_map_update_debugger()` which is hooked into GDB by our integration script.

I find these details fascinating and worth describing, even if very briefly -- it gives at least some hints for interested devs, where to look for corresponding source code.
</blockquote></details>

I added more details and links, but I think `execve` belongs to the "detail overload" category...
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 169 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09r4Ma9vaWJoOwTN1V:-N0AorcnA6a2GVh5CGWv:b-896fix) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L169)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

You have this sentence twice, please remove one copy.
</blockquote></details>

Done.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 171 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09rQ7m6y7pIouE20ep:-N0AumZGDkoC5AkLzA4v:b3gpfh7) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L171)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

In the picture, you have `write(1, "hello")` twice: the upper-most stack frame is for the untrusted runtime, the lower stack frame is for the trusted in-enclave runtime.

I suggest to give some hint, e.g. adding ` (host-level syscall)` and ` (enclave-level syscall)` after each of these two, to distinguish them.
</blockquote></details>

I added this to the reconstructed stack trace (but not to the diagram on the left), is that better?
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 178 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09ryTa9GjkP2QVoQzT:-N0Ap1wNDuTARDfHUbJs:b-b671hy) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L178)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I find this link strange. I thought you wanted to give some real-world examples in this link, or at least point to some GDB official documentation.
</blockquote></details>

Yes, this is a mistake. I had a nice blog post to link here.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (changes requested):
> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @pwmarcz and @woju)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09y1Hp2l_n93lFmIK6:-N0B30kO9TQ8Lv8TlrWy:b2wpqlv) (no related file):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

Thanks. If all goes well, I want to write one or two more (about SGX profiling, and ASan).
</blockquote></details>

Yes, these are very interesting topics. And also good for documentation/future. I hope you'll find the time.

---
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 95 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09uE9l8UVxkkYSTOm4:-N0BAwda3vKgPJ64Gs1O:b-vs1prh) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L95)):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

Yeah, it's a balance - the detail is interesting, but I'm afraid too much will be again boring for the reader. 

I added another paragraph of explanation (and the link to `enclave_dbginfo`), what do you think?
</blockquote></details>

Looks good!
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 171 at r2](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N09rQ7m6y7pIouE20ep:-N0BBVnmAr3pB8k1A721:bb77ufg) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/f502f97aafdea76aef19bd47ce13f4da01ca5da0/content/blog/gdb-support-in-gramine/contents.lr#L171)):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

I added this to the reconstructed stack trace (but not to the diagram on the left), is that better?
</blockquote></details>

Yes.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 77 at r3](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N0BAf0-3qoDBXVcx3am:-N0BAf0-3qoDBXVcx3an:b-endajn) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/d368c778323ac9f6280d4df28021c73cfcdcf507/content/blog/gdb-support-in-gramine/contents.lr#L77)):*
> ```Markdown
> Unfortunately, it's not that easy to check which part of the program is being
> executed. Whenever the process is stopped while in enclave, the SGX hardware
> forces it to exit the eclave and land on the Async Exit Pointer (AEP). Most CPU
> ```

`enclave`


<!-- Sent from Reviewable.io -->

pwmarcz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: 13 of 14 files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv, @pwmarcz, and @woju)

___
*[content/blog/gdb-support-in-gramine/contents.lr, line 77 at r3](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N0BAf0-3qoDBXVcx3am:-N0BMk-P8bJhLZSqbVFb:bt2vr1b) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/d368c778323ac9f6280d4df28021c73cfcdcf507/content/blog/gdb-support-in-gramine/contents.lr#L77)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`enclave`
</blockquote></details>

Fixed.


<!-- Sent from Reviewable.io -->

dimakuv reviewed (dismissed):
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: all files reviewed, 4 unresolved discussions, "fixup! " found in commit messages' one-liners (waiting on @pwmarcz and @woju)



<!-- Sent from Reviewable.io -->

woju reviewed (dismissed):
> Reviewed 2 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: all files reviewed, 2 unresolved discussions, "fixup! " found in commit messages' one-liners (waiting on @pwmarcz)



<!-- Sent from Reviewable.io -->

pwmarcz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: 0 of 14 files reviewed, all discussions resolved, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv and @woju)

___
*[content/blog/gdb-support-in-gramine/contents.lr, line 5 at r1](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N06REx1-7KPTyPsE76r:-N0Bs_ybCQpSpolHbR-I:b-896fix) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/7fa97861dd53ae7a5219185fe6da81b3131919f0/content/blog/gdb-support-in-gramine/contents.lr#L5)):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

TODO on merge: update date
</blockquote></details>

Done.
___
*[content/blog/gdb-support-in-gramine/contents.lr, line 17 at r1](https://reviewable.io/reviews/gramineproject/gramineproject.io/1#-N06RBtyAO5W7--RGsls:-N0Bsaus4n0uzqLk0Azl:b-9mg3uh) ([raw file](https://github.com/gramineproject/gramineproject.io/blob/7fa97861dd53ae7a5219185fe6da81b3131919f0/content/blog/gdb-support-in-gramine/contents.lr#L17)):*
<details><summary><i>Previously, pwmarcz (Paweł Marczewski) wrote…</i></summary><blockquote>

TODO on merge: add reviewers
</blockquote></details>

Done (added @dimakuv).


<!-- Sent from Reviewable.io -->

woju reviewed (approved):
> Reviewed 9 of 14 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: 9 of 14 files reviewed, all discussions resolved, not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv)



<!-- Sent from Reviewable.io -->

dimakuv reviewed (approved):
> Reviewed 14 of 14 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramineproject.io/1)* status: :shipit: complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
